### PR TITLE
Laboratory work #2, Anna Azimkova - 22FPL1 

### PR DESCRIPTION
**Pull request by puglyann**

### CODE REVIEW

**Comment by demid5111** [comment]
temporarily closing it

**Comment by demid5111** [comment]
re-opening

File: lab_2_tokenize_by_bpe/main.py [hunk 1]
```diff
@@ -14,6 +14,23 @@ def prepare_word(
     :param end_of_word: a token that signifies the end of word
     :return: preprocessed word
     """
+    if not (isinstance(raw_word, str)
+            and isinstance(start_of_word, (str, type(None)))
```
Comment (marina-kaz) [inline lab_2_tokenize_by_bpe/main.py:18]: This is a smart workaround, however it is better to check whether the variable is `None` by directly stating `<var> is None` (exactly `is`! can you tell why?)

File: lab_2_tokenize_by_bpe/main.py [hunk 2]
```diff
@@ -14,6 +14,23 @@ def prepare_word(
...
+            and isinstance(end_of_word, str)
+    ):
+        return None
+
+    word_tokens = []
+
+    if start_of_word is not None:
+        word_tokens.append(start_of_word)
+
+    for char in raw_word:
```
Comment (marina-kaz) [inline lab_2_tokenize_by_bpe/main.py:28]: Can we extend the `word_tokens` with the characters of the word without writing a `for` loop?

File: lab_2_tokenize_by_bpe/main.py [hunk 3]
```diff
@@ -26,7 +43,26 @@ def collect_frequencies(
...
+    words = text.split()
+
+    for word in words:
+        prepared_word = prepare_word(word, start_of_word, end_of_word)
+        if prepared_word is None:
+            return None
+
+        if prepared_word in frequencies:
+            frequencies[prepared_word] += 1
+        else:
```
Comment (marina-kaz) [inline lab_2_tokenize_by_bpe/main.py:62]: Can we avoid `else` here?

**Review by marina-kaz — COMMENTED** [review]
Good start! Keep going

**Comment by marina-kaz** [comment]
Congratulation with the best implementation of the 1st lab!

File: lab_2_tokenize_by_bpe/main.py (hunk 2 above)
Comment (marina-kaz) [reply lab_2_tokenize_by_bpe/main.py:29]: Still applies

File: lab_2_tokenize_by_bpe/main.py [hunk 4]
```diff
@@ -26,6 +44,23 @@ def collect_frequencies(
...
+
+    frequencies = {}
+    words = text.split()
+
+    for word in words:
+        prepared_word = prepare_word(word, start_of_word, end_of_word)
+        if prepared_word is None:
+            return None
+
+        frequencies.update({prepared_word: words.count(word)})
```
Comment (marina-kaz) [inline lab_2_tokenize_by_bpe/main.py:61]: `update` exists to add multiple keys in the dictionary at once, if you only need to add one key, consider adding it traditionally with the square brackets

File: lab_2_tokenize_by_bpe/main.py [hunk 5]
```diff
@@ -36,6 +71,23 @@ def count_tokens_pairs(
...
+
+    for word_tokens in word_frequencies:
+        for i in range(len(word_tokens) - 1):
+            token = word_tokens[i]
+            next_token = word_tokens[i + 1]
+            pair = (token, next_token)
+
+            if pair not in pair_frequency_dict:
+                pair_frequency_dict[pair] = word_frequencies[word_tokens]
+            else:
```
Comment (marina-kaz) [inline lab_2_tokenize_by_bpe/main.py:87]: Can we avoid this `else`?

File: lab_2_tokenize_by_bpe/main.py [hunk 6]
```diff
@@ -58,6 +130,29 @@ def train(
...
+        if not pair_frequency_dict:
+            return None
+
+        if num_merges > len(pair_frequency_dict):
+            num_merges = len(pair_frequency_dict)
+
+        max_occurrence = max(pair_frequency_dict.values())
+        good_pairs = [pair for pair, frequency in pair_frequency_dict.items() if frequency == max_occurrence]
+        max_len = max(len(str(pair)) for pair in good_pairs)
+        long_pairs = [pair for pair in good_pairs if len(str(pair)) == max_len]
```
Comment (marina-kaz) [inline lab_2_tokenize_by_bpe/main.py:147]: Consider using `sorted` with double sorting order instead of filtering this way

**Review by marina-kaz — COMMENTED** [review]
Implementations are very good, keep going!

File: lab_2_tokenize_by_bpe/main.py [hunk 7]
```diff
@@ -69,6 +159,24 @@ def get_vocabulary(
...
     :return: dictionary in the form of <token: identifier>
     """
+    if not (
+        isinstance(word_frequencies, dict)
+        and isinstance(unknown_token, str)
+    ):
+        return None
+
+    vocabulary = {}
+    not_good_list = set()
```
Comment (marina-kaz) [inline lab_2_tokenize_by_bpe/main.py:169]: Not good variable name

File: lab_2_tokenize_by_bpe/main.py [hunk 8]
```diff
@@ -81,6 +189,23 @@ def decode(
...
+        and (isinstance(end_of_word_token, str) or end_of_word_token is None)
+    ):
+        return None
+
+    vocabulary_inverted = {id: token for token, id in vocabulary.items()}
+    decoded_text = str()
+    for number in encoded_text:
+        token = vocabulary_inverted[number]
+        if end_of_word_token and end_of_word_token in token:
+            token = token.replace(end_of_word_token, ' ')
```
Comment (marina-kaz) [inline lab_2_tokenize_by_bpe/main.py:205]: Consider replacing it once when all the tokens are appended

**Review by marina-kaz — COMMENTED** [review]
Good progress, keep going!
Not all coments are addressed.

**Review by marina-kaz — COMMENTED** [review]
No changes 😿

File: lab_2_tokenize_by_bpe/main.py [hunk 9]
```diff
@@ -47,6 +94,27 @@ def merge_tokens(
...
+    for word in word_frequencies.keys():
+        if ''.join(pair) in ''.join(word):
+            list_word = list(word)
+            for i in range(len(word) - 1):
+                new_key = (word[i], word[i + 1])
+                if new_key == pair:
+                    list_word[i + 1] = ''.join(pair)
+                    list_word.pop(i)
+            merged_tokens_dict[tuple(list_word)] = word_frequencies[word]
+        else:
```
Comment (marina-kaz) [inline lab_2_tokenize_by_bpe/main.py:114]: This else could have been easily avoided

**Pull request closed without merging**